File: 69.sqrt-x.cpp
/*
 * @lc app=leetcode id=69 lang=cpp
 *
 * [69] Sqrt(x)
 */

// @lc code=start
class Solution {
public:
    int mySqrt(int x) {
        if (x == 0) {
            return 0;
        }
        int s = 1, e = x;
        while (s <= e) {
            int m = s + ((e - s) >> 1);
            if (m > x / m) {
                e = m - 1;
            } else {
                if (m + 1 > x / (m + 1)) {
                    return m;
                }
                s = m + 1;
            }
        }

        return 0;
    }
};
// @lc code=end

File: 69.sqrt-x.rs
/*
 * @lc app=leetcode id=69 lang=rust
 *
 * [69] Sqrt(x)
 */

// @lc code=start
impl Solution {
    pub fn my_sqrt(x: i32) -> i32 {
        let mut s = 0;
        let mut e = x;
        while s <= e {
            let m = s + ((e - s) >> 1);
            let pow = m * m;
            if pow == x {
                return m;
            }
            if pow > x || pow < 0 {
                e -= 1;
            } else if (m + 1) * (m + 1) > x || (m + 1) * (m + 1) < 0 {
                return m;
            } else {
                s += 1;
            }
        }

        0
    }
}
// @lc code=end

